### PR TITLE
Safer Field::safeGet?

### PR DESCRIPTION
**Pull request by yakov-olkhovskiy**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

`Field::safeGet<>()` always returns reference to its internal field which makes it unsafe when `Field` is a temporary - this PR is trying to fix it by providing reference qualified `safeGet` methods.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 76dee30deebaad1e8e7bdb463d10aaa31ee59444 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72659/76dee30deebaad1e8e7bdb463d10aaa31ee59444/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/b81ee27750c4520de6e6b979ce7d4f6a111cb8ea/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Review by evillique — APPROVED** [review]

File: src/Core/Field.h [hunk 1]
```diff
@@ -479,12 +479,20 @@ class Field
...
+    template <typename T> auto safeGet() const &&
+    {
+        return safeGet<T>();
+    }
 
-    template <typename T> auto & safeGet();
+    template <typename T> auto & safeGet() &;
+    template <typename T> auto safeGet() &&
+    {
+        return safeGet<T>();
```
Comment (antonio2368) [inline src/Core/Field.h:494]: `std::move`?

Comment (yakov-olkhovskiy) [reply src/Core/Field.h:494]: exactly!
thanks

Comment (yakov-olkhovskiy) [reply src/Core/Field.h:494]: ...though I'm not sure we use move semantics in any of ours Field types :)
but probably now we should - will look into it on next iteration :)

**Pull request merged**